server/src/main/java/org/opensearch/action/ActionModule.java

```diff
@@ -1071,5 +1095,54 @@ public void unregisterDynamicAction(ActionType<?> action) {
             }
             return registry.get(action);
         }
+
+        /**
+         * Add a dynamic action to the registry.
+         *
+         * @param route The route instance to add
+         * @param action The corresponding instance of RestSendToExtensionAction to execute
+         */
+        public void registerDynamicRoute(RestHandler.Route route, RestSendToExtensionAction action) {
+            requireNonNull(route, "route is required");
+            requireNonNull(action, "action is required");
+            Optional<String> routeName = Optional.empty();
+            if (route instanceof NamedRoute) {
+                routeName = Optional.of(((NamedRoute) route).name());
+                if (isActionRegistered(routeName.get()) || registeredActionNames.contains(routeName.get())) {
+                    throw new IllegalArgumentException("route [" + route + "] already registered");
+                }
+            }
+            if (routeRegistry.containsKey(route)) {
+                throw new IllegalArgumentException("route [" + route + "] already registered");
+            }
+            routeRegistry.put(route, action);
+            routeName.ifPresent(registeredActionNames::add);
+        }
+
+        /**
+         * Remove a dynamic route from the registry.
+         *
+         * @param route The route to remove
+         */
+        public void unregisterDynamicRoute(RestHandler.Route route) {
+            requireNonNull(route, "route is required");
+            if (routeRegistry.remove(route) == null) {
+                throw new IllegalArgumentException("action [" + route + "] was not registered");
+            }
+            if (route instanceof NamedRoute) {
+                registeredActionNames.remove(((NamedRoute) route).name());
+            }
+        }
```

Comment:
Need test for this method.

Reply:
There was a basic test of DynamicActionRegistry within the ActionModuleTests. I created a new test suite called DynamicActionRegistryTests where we can add a lot more tests of different scenarios. Some of the logic is already covered in RestSendToExtensionTests, but I agree that its a good idea to more directly test the DynamicActionRegistry.

I added 2 new tests that directly validate the registerDynamicRoute and unregisterDynamicRoute methods.